### ISSUE TITLE
[FR] Add generic HDF5 dataset tracking

### ISSUE THREAD

**Issue by nnethery** [mlflow/mlflow#25442]
> [!WARNING]
> Before submitting a PR, please make sure that:
> - A maintainer has triaged this issue and applied the `ready` label
> - This issue has no assignee
> - No duplicate PR exists
>
> PRs not meeting these requirements may be automatically closed.

### Willingness to contribute

Yes. I can contribute this feature independently.

### Proposal Summary

Add read-only HDF5 dataset tracking through h5py.
The proposed API is:
``` py
mlflow.data.load_hdf5(
    source,
    *,
    root="/",
    features=None,
    targets=None,
    name=None,
    digest=None,
)
```

It would return an HDF5Dataset containing source lineage, hierarchy-aware schema and profile metadata, and a scalable sampled digest.

An HDF5 file is a container rather than a single table or array. root therefore selects the hierarchy being tracked, while features and targets separately identify arrays used as model inputs or labels.

Tracking would remain reference-only. Logging an HDF5Dataset would not automatically upload the underlying file as an artifact.

#### What is the use case for this feature?
Track HDF5 files used for training or evaluation without first converting them into a DataFrame or loading the entire file into a NumPy array.

A file may contain several groups, arrays, attributes, and links. Users should be able to track the complete file, a subgroup, or one leaf dataset while preserving that structure in MLflow’s dataset metadata.
For evaluation, users should be able to select one or more HDF5 datasets as model inputs and optionally select targets.

#### Why is this use case valuable to support for MLflow users in general?
HDF5 is used for scientific data, image and signal datasets, simulation output, embeddings, and deep-learning pipelines. These files commonly contain several independently named arrays rather than one tabular dataset.
Converting an HDF5 file into a currently supported MLflow dataset type loses useful information:
- Group and dataset paths
- Attributes
- Chunking and compression settings
- Resizable and virtual dataset definitions
- HDF5-specific dtype information
- The relationship between several arrays stored in the same file
Generic HDF5 support would preserve this information while using the existing MLflow Dataset abstraction and tracking backend.
#### Why is this use case valuable to support for your project(s) or organization?
I would like to track .h5 training and evaluation data without flattening the file into a table or treating every contained array as an unrelated dataset.

### Details

The initial implementation would be generic, read-only HDF5 support. It would not interpret Pandas `HDFStore` layouts or Keras model files.

`h5py` would remain optional and be imported lazily. The proposed supported range is `h5py>=3.11,<4`.

`load_hdf5` would accept:
- Local paths
- File URIs
- Remote URIs supported by MLflow dataset sources
- `DatasetSource` instances whose `load() `method resolves to a local file
Remote sources would be downloaded for inspection while retaining the original URI as lineage.
The resolved file would be validated with `h5py.is_hdf5`. HDF5 handles would only remain open during inspection or selected-array reads.

`root="/"` would track the whole file. A group or dataset path could be supplied to track a smaller hierarchy.
Traversal would be deterministic. Groups, datasets, and links would be recorded in lexicographic order. Soft and external links would be described without dereferencing them. Hard-link aliases would use a deterministic canonical target.

The schema would describe:
- Object type and HDF5 path
- Shape and maximum shape
- NumPy and HDF5-specific dtype information
- Chunking and compression
- Fill values
- Virtual and external storage definitions
- Attribute names, shapes, and dtypes
- Hard, soft, and external links
Compound dtypes would be described recursively so nested strings, enums, variable-length values, opaque values, and references remain visible.

The profile would contain hierarchy counts, attribute and element counts, nominal and allocated storage sizes, selector paths, unsupported-value counts, and schema truncation state.

MLflow currently limits dataset schemas to 1 MiB. Large HDF5 schemas would be truncated deterministically while retaining complete counts in the profile.

The digest would include:
- Normalized hierarchy metadata
- Paths, shapes, and dtypes
- Attribute metadata and bounded samples
- Link targets
- Virtual and external storage definitions
- Feature and target selector configuration
Reference-backed values and external content would not be dereferenced. Their definitions would still affect the digest.

This is intentionally a scalable fingerprint rather than a complete checksum. A modification outside the sampled region of a large array may not change the digest, but I think this is acceptable. 

Whole-file tracking would not require selectors.

Using the dataset with model evaluation would require features. A selector could be one HDF5 path or an alias-to-path mapping for named inputs. targets would use the same form.

Selected arrays would be loaded only when accessed or converted for evaluation. Scalars, incompatible dtypes, virtual or externally stored arrays, and mismatched first dimensions would produce actionable errors.

No tracking-store, REST, database, or UI changes should be required. The existing Dataset entity already persists name, digest, source, schema, and profile.

The `.h5` file would not be uploaded automatically. Users who need an immutable copy could log it separately with `mlflow.log_artifact`.

### What machine learning domain(s) is this feature request about?

- [ ] `domain/genai`: LLMs, Agents, and other GenAI-related use cases
- [ ] `domain/classical-ml`: Traditional machine learning, such as linear regression.
- [ ] `domain/deep-learning`: Deep learning and neural networks.
- [x] `domain/platform`: MLflow platform foundation, not specific to a particular machine learning domain.

### What area(s) of MLflow is this feature request about?

- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/scoring`: MLflow model serving, deployment tools, Spark UDFs
- [x] `area/evaluation`: MLflow model evaluation features, evaluation metrics, and evaluation workflows
- [ ] `area/prompt`: MLflow prompt engineering features, prompt templates, and prompt management
- [ ] `area/tracing`: MLflow Tracing features, tracing APIs, and LLM tracing functionality
- [ ] `area/gateway`: MLflow AI Gateway client APIs, server, and third-party integrations
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/uiux`: Front-end, user experience, plotting
- [x] `area/docs`: MLflow documentation pages

**Comment by TomeHirata** [comment]
Marked as ready!

**Comment by luisangelrod** [comment]
Hi, I'd like to work on this. I have a local implementation following the proposed API, including reference-only tracking, bounded digest sampling, selector support, schema and profile limit handling, documentation, and focused tests. I'll keep the PR scoped to #25442.

**Comment by Kunal8954** [comment]
Hi! I'd like to work on this issue. I’ve gone through the proposal and would like to implement the generic read-only HDF5 dataset tracking through h5py, including the associated tests and documentation. Please let me know if there are any design considerations or existing components I should follow before I start.

**Comment by luisangelrod** [comment]
Correction: I had scoped the proposed implementation but do not yet have a complete local implementation. Since the issue is currently assigned and another contributor has expressed interest, I will wait for maintainer coordination rather than duplicate work. Apologies for overstating the local progress in my previous comment.
